File: search/facebook_search.py
"""
facebook_search.py — Facebook / Meta source adapter.

Searches public Facebook posts, groups, and marketplace listings
for Singing Bowls buyers/importers.
NOTE: Facebook heavily restricts scraping; this adapter uses
public search URLs and gracefully handles blocks.
"""

import logging
import re
import time
from typing import List, Dict

import requests
from bs4 import BeautifulSoup

from config import SEARCH_KEYWORD, MAX_RESULTS_PER_SOURCE, IMAGE_EXTENSIONS

logger = logging.getLogger(__name__)

HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (X11; Linux x86_64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/120.0.0.0 Safari/537.36"
    ),
    "Accept-Language": "en-US,en;q=0.9",
}

EMAIL_PATTERN = re.compile(r"[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}")

# Public directories/marketplaces that aggregate Facebook leads
FALLBACK_SOURCES = [
    "https://www.importgenius.com/search?q={keyword}",
    "https://www.tradekey.com/search/?searchType=products&keyword={keyword}+import",
    "https://www.globalsources.com/gsol/I/Singing-bowls/I/product.htm",
]


def _extract_emails_from_text(text: str) -> List[str]:
    candidates = EMAIL_PATTERN.findall(text)
    return [
        e.lower() for e in candidates
        if not any(e.lower().endswith(ext) for ext in IMAGE_EXTENSIONS)
        and len(e.split("@")[-1]) <= 50
    ]


def _try_public_search(keyword: str) -> List[Dict]:
    """Attempt to extract leads from public business directories as a fallback."""
    records = []
    url = f"https://www.tradekey.com/search/?searchType=buyers&keyword={requests.utils.quote(keyword)}"
    try:
        resp = requests.get(url, headers=HEADERS, timeout=10)
        soup = BeautifulSoup(resp.text, "lxml")
        for item in soup.select(".product-item, .buyer-item, .listing")[:10]:
            text = item.get_text(separator=" ", strip=True)
            emails = _extract_emails_from_text(text)
            company = item.select_one("h2, h3, .company-name")
            country = item.select_one(".country, .location")
            if emails:
                records.append({
                    "buyer_name": "",
                    "company_name": company.get_text(strip=True) if company else text[:60],
                    "email": emails[0],
                    "website": url,
                    "country": country.get_text(strip=True) if country else "",
                    "source_platform": "Facebook/TradeKey",
                })
        time.sleep(1)
    except Exception as exc:
        logger.warning("Facebook fallback scrape failed: %s", exc)

    return records


def search(keyword: str = SEARCH_KEYWORD, max_results: int = MAX_RESULTS_PER_SOURCE) -> List[Dict]:
    """
    Main entry point for the Facebook source adapter.
    Returns buyer record dicts.
    """
    logger.info("Facebook adapter: starting search for '%s'", keyword)
    records = _try_public_search(keyword)

    # Try additional fallback sources
    for tmpl in FALLBACK_SOURCES[1:]:
        if len(records) >= max_results:
            break
        try:
            url = tmpl.format(keyword=requests.utils.quote(keyword))
            resp = requests.get(url, headers=HEADERS, timeout=10)
            soup = BeautifulSoup(resp.text, "lxml")
            text = soup.get_text(separator=" ")
            emails = _extract_emails_from_text(text)
            for email in emails[:5]:
                records.append({
                    "buyer_name": "",
                    "company_name": "",
                    "email": email,
                    "website": url,
                    "country": "",
                    "source_platform": "Facebook/Directory",
                })
            time.sleep(1)
        except Exception as exc:
            logger.debug("Fallback source failed: %s", exc)

    logger.info("Facebook adapter: found %d records", len(records))
    return records[:max_results]

File: src/lib/search/facebookSearch.ts
/**
 * facebookSearch.ts — Facebook source adapter (best-effort scraping).
 * Port of search/facebook_search.py
 */

import * as cheerio from 'cheerio';
import { SEARCH_KEYWORD, MAX_RESULTS_PER_SOURCE } from '@/lib/config';
import type { BuyerRecord } from '@/types';

const EMAIL_PATTERN = /[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/g;

const HEADERS = {
  'User-Agent':
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
};

export async function search(
  keyword = SEARCH_KEYWORD,
  maxResults = MAX_RESULTS_PER_SOURCE
): Promise<Partial<BuyerRecord>[]> {
  const results: Partial<BuyerRecord>[] = [];
  try {
    // Facebook heavily resists scraping — do a public Bing search for Facebook pages
    const query = `site:facebook.com "${keyword}" importer OR buyer`;
    const url = `https://www.bing.com/search?q=${encodeURIComponent(query)}&count=10`;
    const resp = await fetch(url, { headers: HEADERS, signal: AbortSignal.timeout(10000) });
    const html = await resp.text();
    const $ = cheerio.load(html);

    $('li.b_algo').each((_, el) => {
      const title = $(el).find('h2').text().trim();
      const snippet = $(el).find('.b_caption p').text().trim();
      const href = $(el).find('a').attr('href') ?? '';
      const emails = (snippet.match(EMAIL_PATTERN) ?? []).map((e) => e.toLowerCase());
      if (emails.length && href) {
        results.push({
          buyer_name: title,
          company_name: title,
          email: emails[0],
          website: href,
          country: '',
          source_platform: 'Facebook',
          discovered_at: new Date().toISOString(),
        });
      }
      if (results.length >= maxResults) return false;
    });
  } catch {
    // best effort
  }
  return results.slice(0, maxResults);
}
